From 67dba97e4598eaf2deb14da044fbfb1c119cf76f Mon Sep 17 00:00:00 2001
From: Andrei Borzenkov <arvidjaar@gmail.com>
Date: Wed, 3 Feb 2016 20:34:55 +0300
Subject: support modules without symbol table

all_video module does not have any code or data and exists solely for
.moddeps section to pull in dependencies. This makes all symbols unneeded.

While in current binutils (last released version as of this commit is 2.26)
``strip --strip-unneeded'' unintentionally adds section symbols for each
existing section, this behavior was considered a bug and changed in commit
14f2c699ddca1e2f706342dffc59a6c7e23e844c to completely strip symbol table
in this case.

Older binutils (verified with 2.17) and some other toolchains (at least
elftoolchain r3223M), both used in FreeBSD, remove symbol table in all_video
as well.

Relax run-time check and do not return error for modules without symbol table.
Add additional checks to module verifier to make sure such modules

a) have non-empty .moddeps section. Without either externally visible symbols
or .moddeps modules are completely useless and should not be built.

b) do not have any relocations.

Closes: 46986

v2: add run-time check for empty symbol table if relocations are present as
    suggested by Vladimir.
---
 grub-core/kern/dl.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

Index: b/grub-core/kern/dl.c
===================================================================
--- a/grub-core/kern/dl.c
+++ b/grub-core/kern/dl.c
@@ -337,8 +337,11 @@ grub_dl_resolve_symbols (grub_dl_t mod,
     if (s->sh_type == SHT_SYMTAB)
       break;
 
+  /* Module without symbol table may still be used to pull in dependencies.
+     We verify at build time that such modules do not contain any relocations
+     that may reference symbol table. */
   if (i == e->e_shnum)
-    return grub_error (GRUB_ERR_BAD_MODULE, N_("no symbol table"));
+    return GRUB_ERR_NONE;
 
 #ifdef GRUB_MODULES_MACHINE_READONLY
   mod->symtab = grub_malloc (s->sh_size);
@@ -580,6 +583,9 @@ grub_dl_relocate_symbols (grub_dl_t mod,
 
 	if (seg)
 	  {
+	    if (!mod->symtab)
+	      return grub_error (GRUB_ERR_BAD_MODULE, "relocation without symbol table");
+
 	    err = grub_arch_dl_relocate_symbols (mod, ehdr, s, seg);
 	    if (err)
 	      return err;
